### ISSUE TITLE
BUG: CUDA (cudaMemset) error

### ISSUE THREAD

**Issue by joabro** [brucefan1983/GPUMD#46]
What happens?
---------------

When running GPUMD the following error, sometimes, occurs:
```
    CUDA Error:
    File:       neighbor_ON1.cu
    Line:       222
    Error code: 1
    Error text: invalid argument
```

What is the reason?
--------------------
As is indicated by the error message, it is related to line 222 in [`neighbor_ON1.cu`](GPUMD/src/neighbor_ON1.cu):
```
CHECK(cudaMemset(cell_count, 0, sizeof(int)*N_cells));
```

What is the expected behaviour?
--------------------------------
It is possible to run GPUMD on systems with multiple GPUs without receiving any error messages.

**Comment by joabro** [comment]
cc @brucefan1983

**Comment by brucefan1983** [comment]
GPUMD does will choose to run on the availble GPU, starting from the
smallest device ID. If you want to choose the GPU with an ID of GPU_ID, you
can use

$ export CUDA_VISIBLE_DEVICES=GPU_ID

before submitting your GPUMD job.

BTW, why you need to use cudaSetDevice? Are you integrating GPUMD into some
other program?


joabro <notifications@github.com> 于 2020年1月13日周一 17:00写道：

**Comment by joabro** [comment]
I do not want to use cudaSetDevice, I just found this thread when I googled the problem on the internet. I am sorry that I did not make this clear, but the two blocks of text under "What is the reason" are quotes from that thread, so the words are not mine. I just thought that this might help you to solve the problem.

Anyhow, I will try your suggestion and select the device ID, but this might be difficult since I do not know which of the two GPUs that I have been allocated. I think I will have forward this question to the support staff for the computer cluster so that they can tell me how to write the submit scripts. I guess the worst case scenario is that I would need to request both GPUs and then run the simulation on only one of them, which has the drawback that the calculations will remain longer in the queue and require double as many core hours.
________________________________

**Comment by brucefan1983** [comment]
You can try to run the command

$ nvidia-smi

to check the IDs of the GPUs and what GPUs are busy.

Apart from selecting the GPU for use before running GPUMD, you can also
modify the code: just adding
cudaSetDevice(your_gpu_id);
in the begining of the main function in main.cu.

Zheyong

joabro <notifications@github.com> 于 2020年1月13日周一 18:35写道：

**Comment by joabro** [comment]
@brucefan1983 I have just found out that this problem occurs  even though I am running the simulation on a node with a single GPU. I will change the title and description accordingly.

**Comment by brucefan1983** [comment]
Could you send me the input files? 2nd order is enough.

**Comment by joabro** [comment]
Here are the input files:
[test_vel100_nvt_lan_T100_Tc200_dump1000_run100000.tar.gz](https://github.com/brucefan1983/GPUMD/files/4055612/test_vel100_nvt_lan_T100_Tc200_dump1000_run100000.tar.gz)
As well as the potential:
[ols_original_model_1_2x2x2_orders_2-3_tol1e-08.tar.gz](https://github.com/brucefan1983/GPUMD/files/4055614/ols_original_model_1_2x2x2_orders_2-3_tol1e-08.tar.gz)

**Comment by brucefan1983** [comment]
Thanks for your inputs. I have solved the problem. It is bug in the code related to the neighbor list construction/update. Some arrays only have a length of `N` (number of atoms), but when the cutoff distance is 0.1 Angstrom (I recommended this to you, sorry), the number of bins exceeds `N`, which results in some invalid memory access. For the time being, I have disabled neighbor list construction/update for the FCP potential (need `-DUSE_FCP` in the `makefile`), becasue we don't need the neighbor list at all for this potential. 

For your inputs, I would like to mention that, to use netdcf, you need to build it according to the manual:

https://gpumd.zheyongfan.org/index.php/NetCDF_setup
https://gpumd.zheyongfan.org/index.php/The_dump_position_keyword

Without the compute command, your inputs run correltly in my laptop now. 

Zheyong

**Comment by brucefan1983** [comment]
Sorry, I have a typo, 
Without the compute command
should be
Without the dump_position command

**Comment by joabro** [comment]
Thank for the taking care of this so quickly @brucefan1983. I have now been able to run all the calculations without any issues.

By the way, I have already installed netcdf according to the instructions that you refer to. It would, however, be great if GPUMD could have support for other, especially more recent, versions of netcdf. Is this something that you are working on?

I suppose it should be possible to close this issue, right?
